test/ex_doc/doc_ast_test.exs

```diff
@@ -0,0 +1,85 @@
+defmodule ExDoc.DocASTTest do
+  use ExUnit.Case, async: true
+  alias ExDoc.DocAST
+
+  describe "parse!/3" do
+    test "markdown" do
+      markdown = """
+      **foo**
+      """
+
+      assert DocAST.parse!(markdown, "text/markdown") ==
+               [{:p, [], [{:strong, [], ["foo"], %{}}], %{}}]
+    end
+
+    test "markdown errors" do
+      assert ExUnit.CaptureIO.capture_io(:stderr, fn ->
+               assert DocAST.parse!("`String.upcase", "text/markdown") ==
+                        [{:p, [], ["`String.upcase"], %{}}]
+             end) =~ "(warning) nofile:1 Closing unclosed backquotes ` at end of input"
+    end
+
+    test "erlang+html" do
+      erl_ast = [{:p, [], ["foo"]}]
+
+      assert DocAST.parse!(erl_ast, "application/erlang+html") ==
+               [{:p, [], ["foo"], %{}}]
+    end
+  end
+
+  describe "to_string/2" do
+    test "simple" do
+      markdown = """
+      foo **bar** baz
+      """
+
+      ast = DocAST.parse!(markdown, "text/markdown")
+
+      assert DocAST.to_string(ast) ==
+               ~s{<p>foo <strong>bar</strong> baz</p>}
+    end
+
+    test "escape" do
+      markdown = "foo > bar"
+      ast = DocAST.parse!(markdown, "text/markdown")
+      assert DocAST.to_string(ast) == ~s{<p>foo &gt; bar</p>}
+    end
+
+    test "raw html" do
+      markdown = "<strong><em>bar</em></strong>"
+      ast = DocAST.parse!(markdown, "text/markdown")
+      assert DocAST.to_string(ast) == ~s{<strong><em>bar</em></strong>}
+    end
+
+    test "empty" do
+      markdown = """
+      <span>
+      <i></i>
+      <i></i>
+      </span>
+      """
+
+      ast = DocAST.parse!(markdown, "text/markdown")
+      assert DocAST.to_string(ast) == ~s{<span><i></i><i></i></span>}
+    end
+
+    test "with fun" do
+      markdown = """
+      foo **bar** baz
+      """
+
+      ast = DocAST.parse!(markdown, "text/markdown")
+
+      f = fn
+        {:strong, _, _, _}, string ->
+          String.upcase(string)
```

Comment:
note for posterity: this is basically how we could do auto-linking, we just look for `<code>` nodes, see if they are valid refs, and return `<a href="#{url}">#{string}</a>`!

The only problem is this to_string/2 algorithm is a "postwalk" traversal. It is totally fine for regular code blocks but not for custom links, e.g. we'd unnecessarily first render `` `Bar` `` in `` [`Foo`](`Bar`) ``. So we may not end up using this function in which case I'll remove it later, or maybe there's a prewalk approach to consider. Or we add some special handling for custom links, though I'd rather avoid that. But yeah, I decided to keep this function for now as I believe it's really cool we can do it this way :)